base_layer/wallet/src/text_message_service.rs

```diff
@@ -436,3 +608,69 @@ impl TextMessageServiceApi {
         res
     }
 }
+
+mod test {
+    use crate::text_message_service::{Contact, TextMessageError, TextMessageService};
+    use tari_comms::types::{CommsPublicKey, CommsSecretKey};
+    use tari_crypto::keys::{PublicKey, SecretKey};
+
+    #[test]
+    fn test_contacts_crud() {
+        let mut rng = rand::OsRng::new().unwrap();
+
+        let secret_key = CommsSecretKey::random(&mut rng);
+        let public_key = CommsPublicKey::from_secret_key(&secret_key);
+
+        let mut tms = TextMessageService::new(public_key);
+
+        let mut contacts: Vec<Contact> = Vec::new();
+
+        let screen_names = vec![
+            "Alice".to_string(),
+            "Bob".to_string(),
+            "Carol".to_string(),
+            "Dave".to_string(),
+            "Eric".to_string(),
+        ];
+        for i in 0..5 {
+            let contact_secret_key = CommsSecretKey::random(&mut rng);
+            let contact_public_key = CommsPublicKey::from_secret_key(&contact_secret_key);
+            contacts.push(Contact {
+                screen_name: screen_names[i].clone(),
+                pub_key: contact_public_key,
+                address: "127.0.0.1:12345".parse().unwrap(),
+            });
+        }
+
+        assert_eq!(tms.get_screen_name(), None);
+        tms.set_screen_name("Fred".to_string());
+        assert_eq!(tms.get_screen_name(), Some("Fred".to_string()));
+
+        for c in contacts.iter() {
+            tms.add_contact(c.clone())
+        }
+
+        assert_eq!(tms.get_contacts().len(), 5);
+
+        tms.remove_contact(contacts[0].clone()).unwrap();
+
+        assert_eq!(tms.get_contacts().len(), 4);
+
+        contacts[1].screen_name = "Betty".to_string();
+
+        tms.update_contact(&contacts[1]).unwrap();
+
+        let updated_contacts = tms.get_contacts();
+        assert_eq!(updated_contacts[0].screen_name, "Betty".to_string());
+
+        match tms.update_contact(&Contact {
```

Comment:
Perhaps `update_contact` should take an `UpdateContact` struct which does not have a public key

Reply:
You need to search by Public Key though? So I could split it up in the parameters but feels neater to just use the existing struct for all the data?

Reply:
The reasoning is if you should never be allowed to update the public key (assume you'd rather just add another contact and delete the old one) that should not be available as a field for updates. Quite a common pattern for type-safe CRUD stuff - if you agree and it's a quick change then I think go for it, otherwise we can allow time to think about it and change it if needed

Reply:
Searching should perhaps happen (if possible in grpc) by passing in the public key - makes no sense to search with a full Contact object, but I could be misunderstanding